File: libs/csharp/GPBinFiles.cs
//
// Body of GPBinFiles interface.
// This file implements the code of the GPBinFiles.cp file.
// dwc August 1999. COOL version kjg May 2000
// kjg September 2000. Renamed from GPFiles to GPBinFiles.
// kjg March 2001.  Version for Beta-2 libraries.
// 
// Compile with: csc /t:library /r:GPFiles.dll GPBinFiles.cs
//
/* ------------------------------------------------------------ */
using System;
using GPFiles;

namespace GPBinFiles {
public class GPBinFiles {

  /* ----------------------------------	*/

  private static System.String mkStr(char[] arr) {
    int  ix = 0;
    char ch;
    do {
	ch = arr[ix]; ix++;
    } while (ch != '\0');
    return new System.String(arr,0,ix-1);
  }

  /* ----------------------------------	*/

  public static int length(FILE cpf) {
    return (int) cpf.bufS.Length;
  }

  /* ----------------------------------	*/

  private static FILE open(System.String name) {
  // Opens buffered filestream for reading.
    try {
	FILE cpf = new FILE();
	cpf.path = name;
	System.IO.FileStream fStr = 
			System.IO.File.Open(name, System.IO.FileMode.Open);
	cpf.bufS = new System.IO.BufferedStream(fStr);
	return cpf;
    } catch {
	return null;
    }
  }

  private static FILE openRead(System.String name) {
  // Opens buffered filestream for reading.
    try {
	FILE cpf = new FILE();
	cpf.path = name;
	System.IO.FileStream fStr = System.IO.File.OpenRead(name);
	cpf.bufS = new System.IO.BufferedStream(fStr);
	return cpf;
    } catch {
	return null;
    }
  }

 /* =========================== */
                               
  /* ----------------------------------	*/

  public static FILE findLocal(char[] fileName) 
  {
    System.String name = mkStr(fileName);
    return open(name);
  }
                               
  /* ----------------------------------	*/

  public static FILE findOnPath(
			char[] pathName, 
			char[] fileName) 
  { 
    //
    // Use mkStr, to trim space from end of char arrray.
    //
    System.String pName = mkStr(pathName);
    System.String fName = mkStr(fileName);
    System.String nName = "";

    System.String nextDir;
    System.String thisPath = System.Environment.GetEnvironmentVariable(pName);
//
//  System.Console.WriteLine(pName);
//  System.Console.WriteLine(thisPath);
//
    FILE cpf = new FILE();
    bool found = false; 
    bool pathFinished = false;
    int length = thisPath.Length;
    int nextLength;
    int nextPathStart;
    int nextPathEnd   = -1;

    while (!found && !pathFinished) {
      nextPathStart = nextPathEnd + 1;
      nextPathEnd   = thisPath.IndexOf(GPFiles.GPFiles.pathSep, nextPathStart);
      if (nextPathEnd < 0)
	  nextPathEnd = length;
      nextLength    = nextPathEnd - nextPathStart;
      nextDir = thisPath.Substring(nextPathStart, nextLength);
      nName = nextDir + GPFiles.GPFiles.fileSep + fName;
      found = System.IO.File.Exists(nName);
      pathFinished = nextPathEnd >= length; 
    } 
    if (found) {
        return openRead(nName);
    } else
        return null;
  }
   
  /* ----------------------------------	*/

  public static char[] getFullPathName(FILE cpf) {
    return cpf.path.ToCharArray(); // not really correct!
  }

  /* ----------------------------------	*/

  public static FILE openFile(char[] fileName) {
    System.String name = mkStr(fileName);
    return open(name);
  }

  public static FILE openFileRO(char[] fileName) {
    System.String name = mkStr(fileName);
    return openRead(name);
  }
                               
  /* ----------------------------------	*/

  public static void CloseFile(FILE cpf) {
    if (cpf.bufS != null)
	cpf.bufS.Close();
  }

  /* ----------------------------------	*/

  public static FILE createFile(char[] arr) {
    FILE cpf = new FILE();
    try {
	System.String name = mkStr(arr);
	System.IO.FileStream fStr = System.IO.File.Create(name);
      cpf.path = name;
	cpf.bufS = new System.IO.BufferedStream(fStr);
	return cpf;
    } catch {
	return null;
    }
  } 

  /* ----------------------------------	*/

  public static FILE createPath(char[] fileName)
  {
    System.String fName = mkStr(fileName);
    try {
	int ix = fName.LastIndexOf(GPFiles.GPFiles.fileSep);
	if (ix > 0) {
        System.String path = fName.Substring(0,ix);
        // Check if exists first?
	  if (!System.IO.Directory.Exists(path)) {
	    System.IO.DirectoryInfo junk = System.IO.Directory.CreateDirectory(path);
        }
	}
	FILE cpf = new FILE();
      cpf.path = fName;
	System.IO.FileStream fStr = System.IO.File.Create(fName);
	cpf.bufS = new System.IO.BufferedStream(fStr);
	return cpf;
    } catch {
	return null;
    }
  }

  /* ----------------------------------	*/

  public static bool EOF(FILE cpf) {
    return cpf.bufS.Position >= cpf.bufS.Length;
  }

  /* ----------------------------------	*/

  public static int readByte(FILE cpf) {
    if (cpf.bufS != null)
	return (int) cpf.bufS.ReadByte();
    else
	throw new System.Exception("File not open for reading");
  } 

  /* ----------------------------------	*/

  public static int  readNBytes(FILE cpf, 
				byte[] buff, 
				int    want)
  {
    if (cpf.bufS != null)
	return cpf.bufS.Read(buff, 0, want);
    else
	throw new System.Exception("File not open for reading");
  } 

  /* ----------------------------------	*/

  public static void WriteByte(FILE cpf, int b) {
    if (cpf.bufS != null)
	cpf.bufS.WriteByte((byte) b);
    else
	throw new System.Exception("File not open for reading");
  } 

  /* ----------------------------------	*/

  public static void WriteNBytes(FILE cpf,
				 byte[] buff,
				 int    want)
  {
    if (cpf.bufS != null)
	cpf.bufS.Write(buff, 0, want);
    else
	throw new System.Exception("File not open for reading");
  } 
}	// end of class GPBinFiles

/* ------------------------------------------------------------ */
/*		File-descriptor for GPBinFiles			*/
/* ------------------------------------------------------------ */

  public class FILE : GPFiles.FILE
  {
     public System.IO.BufferedStream bufS;
  }	// end of class FILE

/* ------------------------------------------------------------ */
} // end of GPBinFiles.
/* ------------------------------------------------------------ */

File: libs/java/GPBinFiles.java
//
// Body of GPFiles interface.
// This file implements the code of the GPFiles.cp file.
// dwc August 1999.


package CP.GPBinFiles;

import java.io.*;
import CP.CPJ.CPJ;
import CP.GPFiles.GPFiles.*;

public class GPBinFiles {

  public static int length(GPBinFiles_FILE cpf) {
    return (int) cpf.length;
  }

  public static GPBinFiles_FILE findLocal(char[] fileName) 
                                                  throws IOException {
    String currDir = System.getProperty("user.dir");
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    cpf.f = new File(currDir, CP.CPJ.CPJ.MkStr(fileName));
    if (!cpf.f.exists()) {
      return null; 
    } else {
      cpf.rf = new RandomAccessFile(cpf.f,"r");
      cpf.length = cpf.rf.length();
      return cpf;
    }
  }
                               
  public static GPBinFiles_FILE findOnPath(char[] pathName, 
                                   char[] fileName) throws IOException { 
    //
    // Use MkStr, to trim space from end of char arrray.
    //
    String pName = CP.CPJ.CPJ.MkStr(pathName);
    String fName = CP.CPJ.CPJ.MkStr(fileName);

    String nextDir;
    String thisPath = System.getProperty(pName);
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    boolean found = false; 
    boolean pathFinished = false;
    int length = thisPath.length();
    int nextPathStart = -1, nextPathEnd = -1;

    while (!found && !pathFinished) {
      nextPathStart = nextPathEnd + 1;
      nextPathEnd = thisPath.indexOf(CP.GPFiles.GPFiles.pathSep,nextPathStart);
      if (nextPathEnd < 0)
	  nextPathEnd = length;
      nextDir = thisPath.substring(nextPathStart,nextPathEnd);
      cpf.f = new File(nextDir,fName);
      found = cpf.f.exists();
      pathFinished = nextPathEnd >= length; 
    } 
    if (found) {
      cpf.rf = new RandomAccessFile(cpf.f,"r");
      cpf.length = cpf.rf.length();
      return cpf;
    } else {
      return null;
    }
  }
    
  public static char[] getFullPathName(GPBinFiles_FILE cpf) {
    return cpf.f.getPath().toCharArray();
  }

  public static GPBinFiles_FILE openFile(char[] fileName)throws IOException{
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    cpf.f = new File(CP.CPJ.CPJ.MkStr(fileName));
    if (!cpf.f.exists()) {
      return null;
    } else {
      cpf.rf = new RandomAccessFile(cpf.f,"rw");
      cpf.length = cpf.rf.length();
      return cpf;
    }
  }

  public static GPBinFiles_FILE openFileRO(char[] fileName)throws IOException{
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    cpf.f = new File(CP.CPJ.CPJ.MkStr(fileName));
    if (!cpf.f.exists()) {
      return null;
    } else {
      cpf.rf = new RandomAccessFile(cpf.f,"r");
      cpf.length = cpf.rf.length();
      return cpf;
    }
  }

  public static void CloseFile(GPBinFiles_FILE cpf) throws IOException {
    cpf.rf.close(); 
  }

  public static GPBinFiles_FILE createFile(char[] fileName)throws IOException {
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    cpf.f = new File(CP.CPJ.CPJ.MkStr(fileName));
    cpf.rf = new RandomAccessFile(cpf.f,"rw");
    cpf.rf.setLength(0);
    cpf.length = 0;
    //  cpf.length = cpf.rf.length();
    return cpf;
  } 

  public static GPBinFiles_FILE createPath(char[] fileName)throws IOException {
    String fName = CP.CPJ.CPJ.MkStr(fileName);
    int ix = fName.lastIndexOf(File.separatorChar);
    if (ix > 0) {
      File path = new File(fName.substring(0,ix));
      if (!path.exists()) { boolean ok = path.mkdirs(); }
    } 
    GPBinFiles_FILE cpf = new GPBinFiles_FILE();
    cpf.f = new File(fName);
    cpf.rf = new RandomAccessFile(cpf.f,"rw");
    cpf.rf.setLength(0);
    cpf.length = 0;
    //    cpf.length = cpf.rf.length();
    return cpf;
  } 

  public static boolean EOF(GPBinFiles_FILE cpf) throws IOException {
    return cpf.rf.getFilePointer() >= cpf.length;
  }

  public static int readByte(GPBinFiles_FILE cpf) throws IOException {
    return cpf.rf.readUnsignedByte();
  } 

  public static int readNBytes(GPBinFiles_FILE cpf, byte[] buff, 
                               int numBytes) throws IOException {
    return cpf.rf.read(buff,0,numBytes);
  } 

  public static void WriteByte(GPBinFiles_FILE cpf,int b) throws IOException{
    cpf.rf.write(b);
  } 

  public static void WriteNBytes(GPBinFiles_FILE cpf,byte[] buff,
                                 int numBytes) throws IOException {
    cpf.rf.write(buff,0,numBytes);
  } 

  
}
